### PR TITLE
Changed rate limit error message string

### PR DESCRIPTION
**Pull request by korovaisdead**
#622 
Old message: ... rate reset in 26m8.032328588s

New messages:
... [rate reset in 12h26m00s]
... [rate reset in 26m08s]
... [rate reset in 8s]

### CODE REVIEW

**Comment by shurcooL** [comment]
I think this looks like a reasonable solution for #622. Thanks for working on this @korovaisdead.

I have some suggestions on how we can improve the code, I'll post a review.

File: github/github.go [hunk 1]
```diff
@@ -880,6 +880,23 @@ func cloneRequest(r *http.Request) *http.Request {
...
 
+func formatDuration(d time.Duration) string {
+	secondsTotal := int(d.Seconds())
+	hours := int(secondsTotal / 60 / 60)
+	minutes := int((secondsTotal - hours*60*60) / 60)
+	seconds := int(secondsTotal - hours*60*60 - minutes*60)
+
+	if hours > 0 {
+		return fmt.Sprintf("%dh%02dm%02ds", hours, minutes, seconds)
+	}
```
Comment (shurcooL) [inline github/github.go]: As far as I know, GitHub rate limit resets at most 1 hour away from current time. So, I think we can get rid of this case for `hours > 0` because it'll never be true.

Even if it's true in some rare case, it's fine for us to print that rate limit is going to be reset in 120 minutes.

File: github/github.go [hunk 2]
```diff
@@ -880,6 +880,23 @@ func cloneRequest(r *http.Request) *http.Request {
 	return r2
 }
 
+func formatDuration(d time.Duration) string {
```
Comment (shurcooL) [inline github/github.go]: formatDuration could use a comment describing what it's meant to do.

```Go
// formatDuration formats d to look like "26m08s" or "8s".
```

Without a comment saying what the function does, one can only infer it from the implementation, and that makes it hard/impossible to ever improve the implementation, or figure out if a certain behavior is a bug or intended.

Comment (shurcooL) [inline github/github.go]: Also, what will `formatDuration` return on negative durations? Does it handle them properly?

It's possible for a `RateLimitError` to be stored for some time, and its `Error` method can be called after `r.Rate.Reset` is in the past. Then, `r.Rate.Reset.Time.Sub(time.Now())` will be a negative duration, so `formatDuration` needs to be able to handle it.

Perhaps it's a good idea to add a few test cases for `formatDuration`, including a negative duration.

Comment (korovaisdead) [inline github/github.go]: Will take a look into it. Thank for sharing the knowledge about the negative values. I've seen them in the test. But thought it's some kind of obsolete. 
So I'll improve the code to support negative numbers + will write some tests. 

Is it possible to assign the issue to me?

Comment (shurcooL) [inline github/github.go]: Sounds good, thanks.

> Is it possible to assign the issue to me?

@gmlewis would have to invite you to org, otherwise we can't do it because of weird GitHub limitations (/cc @kytrinyx FYI, this isn't very friendly to open source projects):

image

I'll just post a comment in the issue to point to this PR, so people will be aware that you're working on it here.

Comment (gmlewis) [inline github/github.go]: Sending invite.

Comment (korovaisdead) [inline github/github.go]: I'm in. You can assign me now. Thanks!

File: github/github.go [hunk 3]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
 	return r2
 }
 
+// formatRateLimitResetDuration formats d to look like "[rate reset in 2s]" or
+// "[rate reset in 87m02s]" for the positive durations. And like "[rate limit was reset 87m02s ago]"
+// for the negative cases.
+func formatRateLimitResetDuration(d time.Duration) string {
```
Comment (shurcooL) [inline github/github.go]: It's good that the func name is descriptive and specific to the task (since it's only being used in one place), but I think we can shorten it to just `formatRateReset`.

It happens to accept a `time.Duration` parameter, but that doesn't mean we need to put "Duration" in its name. Rate reset should be descriptive enough, since it's being passed `r.Rate.Reset` and it's what we call it in the string.

File: github/github.go [hunk 4]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
...
+	isNegative := d < 0
+	var timeString string
```
Comment (shurcooL) [inline github/github.go]: In general, it's a good practice not to do unnecessary things, which includes declaring variables way earlier than necessary.

In this case, `var timeString string` is declared here, but not used/needed until 7 lines below.

Let's move it so it's declared just before it's used, before `if minutes > 0` line.

That should improve readability of this function.

File: github/github_test.go [hunk 5]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
...
+
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
+	}
+
+	d = 14*time.Minute + 2*time.Second
+	got = formatRateLimitResetDuration(d)
+	want = "[rate reset in 14m02s]"
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
```
Comment (shurcooL) [inline github/github_test.go]: Here too.

File: github/github_test.go [hunk 6]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
...
 	}
 }
+
+func TestFormatRateLimitResetDuration(t *testing.T) {
+	d := 120*time.Minute + 12*time.Second
+	got := formatRateLimitResetDuration(d)
+	want := "[rate reset in 120m12s]"
+
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
```
Comment (shurcooL) [inline github/github_test.go]: To be consistent with the most prevalent style, the order should always be "got", then "want". That makes test output more readable. So make this:

```Go
t.Errorf("Format is wrong. got: %v, want: %v", got, want)
```

And in other lines as well.

File: github/github_test.go [hunk 7]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
...
+	want = "[rate reset in 14m02s]"
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
+	}
+
+	d = 2*time.Minute + 2*time.Second
+	got = formatRateLimitResetDuration(d)
+	want = "[rate reset in 2m02s]"
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
```
Comment (shurcooL) [inline github/github_test.go]: Here too.

File: github/github_test.go [hunk 8]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
...
+	}
+
+	d = 12 * time.Second
+	got = formatRateLimitResetDuration(d)
+	want = "[rate reset in 12s]"
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
```
Comment (shurcooL) [inline github/github_test.go]: Here too.

File: github/github_test.go [hunk 9]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
...
+	}
+
+	d = -1 * (2*time.Hour + 2*time.Second)
+	got = formatRateLimitResetDuration(d)
+	want = "[rate limit was reset 120m02s ago]"
+	if got != want {
+		t.Errorf("Format is wrong. want: %v, got: %v", want, got)
```
Comment (shurcooL) [inline github/github_test.go]: Here too.

Comment (korovaisdead) [inline github/github_test.go]: Thank you so much for noticing that. 
Just came from the another language world where all the time I've seen "Expected: bla-bla, but got: bla-bla-bla".

Improved my code :)

File: github/github_test.go [hunk 10]
```diff
@@ -911,3 +911,41 @@ func TestBasicAuthTransport_transport(t *testing.T) {
 		t.Errorf("Expected custom transport to be used.")
 	}
 }
+
+func TestFormatRateLimitResetDuration(t *testing.T) {
+	d := 120*time.Minute + 12*time.Second
+	got := formatRateReset(d)
+	want := "[rate reset in 120m12s]"
+
```
Comment (shurcooL) [inline github/github_test.go]: Nitpick, remove this blank line, it's not needed.

**Comment by shurcooL** [comment]
@gmlewis, how does this look to you?

File: github/github.go [hunk 11]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
...
 
+// formatRateLimitResetDuration formats d to look like "[rate reset in 2s]" or
+// "[rate reset in 87m02s]" for the positive durations. And like "[rate limit was reset 87m02s ago]"
+// for the negative cases.
+func formatRateReset(d time.Duration) string {
+	isNegative := d < 0
+	if isNegative {
+		d *= -1
+	}
+	secondsTotal := int(d.Seconds())
```
Comment (gmlewis) [inline github/github.go]: nit: rounding might be nice, here:

`secondsTotal := int(0.5 + d.Seconds())`

File: github/github.go [hunk 12]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
...
+	minutes := int((secondsTotal) / 60)
```
Comment (gmlewis) [inline github/github.go]: It seems like there is unnecessary casting here.  How about just:

`minutes := secondsTotal / 60`

File: github/github.go [hunk 13]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
...
+	seconds := int(secondsTotal - minutes*60)
```
Comment (gmlewis) [inline github/github.go]: Same here:

`seconds := secondsTotal - minutes*60`

File: github/github.go [hunk 14]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
 	return r2
 }
 
+// formatRateLimitResetDuration formats d to look like "[rate reset in 2s]" or
+// "[rate reset in 87m02s]" for the positive durations. And like "[rate limit was reset 87m02s ago]"
+// for the negative cases.
+func formatRateReset(d time.Duration) string {
+	isNegative := d < 0
```
Comment (gmlewis) [inline github/github.go]: One thought hit me that it might make this simpler and easier to read if `isNegative` was checked only once (and then didn't actually need to be created at all...  something like:

```go
fmtStr := "[rate reset in %v]"
if d < 0 {
	d = -d
	fmtStr = "[rate limit was reset %v ago]"
}
secondsTotal := int(0.5 + d.Seconds())
minutes := secondsTotal / 60
seconds := secondsTotal - minutes*60

var timeString string
if minutes > 0 {
	timeString = fmt.Sprintf("%dm%02ds", minutes, seconds)
} else {
	timeString = fmt.Sprintf("%ds", seconds)
}

return fmt.Sprintf(fmtStr, timeString)
```

Comment (shurcooL) [inline github/github.go]: > Please let me know if you disagree.

In my personal opinion, I find code that uses variables for format string (to printf calls) harder to read than code that uses separate paths and constant format strings. I find having a `isNegative` boolean which does what it says more clear.

When looking at `return fmt.Sprintf(fmtStr, timeString)`, it's hard to immediately see what it'll print (without having to execute an entire block of code in your head). The following is much more clear to me:

```Go
if isNegative {
	return fmt.Sprintf("[rate limit was reset %v ago]", timeString)
}
return fmt.Sprintf("[rate reset in %v]", timeString)
```

File: github/github.go (hunk 12 above)
Comment (shurcooL) [inline github/github.go]: Yep, this is much better.

[`unconvert`](https://github.com/mdempsky/unconvert) tool by @mdempsky helps finds opportunities like this.

File: github/github.go (hunk 14 above)
Comment (gmlewis) [inline github/github.go]: OK, SGTM.

Comment (korovaisdead) [inline github/github.go]: Totally agree with @shurcooL here. Much more easier to read.

File: github/github.go [hunk 15]
```diff
@@ -888,9 +888,9 @@ func formatRateReset(d time.Duration) string {
 	if isNegative {
 		d *= -1
 	}
-	secondsTotal := int(d.Seconds())
-	minutes := int((secondsTotal) / 60)
-	seconds := int(secondsTotal - minutes*60)
+	secondsTotal := int(0.5 + d.Seconds())
+	minutes := (secondsTotal) / 60
```
Comment (gmlewis) [inline github/github.go]: Please remove the unnecessary parens.

Comment (korovaisdead) [inline github/github.go]: Already. You did the review much faster than I typed `git commit --amend` in terminal.

**Comment by korovaisdead** [comment]
@shurcooL Any updates here ? :)

File: github/github.go [hunk 16]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
 	return r2
 }
 
+// formatRateLimitResetDuration formats d to look like "[rate reset in 2s]" or
```
Comment (shurcooL) [inline github/github.go]: Since you've renamed the function, this comment should also be updated to begin with `// formatRateReset formats d to ...`.

File: github/github.go [hunk 17]
```diff
@@ -880,6 +880,32 @@ func cloneRequest(r *http.Request) *http.Request {
...
+	if minutes > 0 {
+		timeString = fmt.Sprintf("%dm%02ds", minutes, seconds)
+	} else {
+		timeString = fmt.Sprintf("%ds", seconds)
+	}
+
+	if isNegative {
+		return fmt.Sprintf("[rate limit was reset %v ago]", timeString)
+	}
+
```
Comment (shurcooL) [inline github/github.go]: I'd suggest removing this blank line, it does not aid readability. That way, both return statements are closer to each other.

File: github/github_test.go [hunk 18]
```diff
@@ -911,3 +911,40 @@ func TestBasicAuthTransport_transport(t *testing.T) {
 		t.Errorf("Expected custom transport to be used.")
 	}
 }
+
+func TestFormatRateLimitResetDuration(t *testing.T) {
```
Comment (shurcooL) [inline github/github_test.go]: Also rename this to `TestFormatRateReset` to match the new function name.

**Comment by korovaisdead** [comment]
Done @shurcooL
